src/main/java/software/amazon/smithy/lsp/SmithyLanguageServer.java

```diff
@@ -417,46 +413,72 @@ public void didChangeWatchedFiles(DidChangeWatchedFilesParams params) {
         LOGGER.info("DidChangeWatchedFiles");
         // Smithy files were added or deleted to watched sources/imports (specified by smithy-build.json),
         // or the smithy-build.json itself was changed
-        Set<String> createdSmithyFiles = new HashSet<>(params.getChanges().size());
-        Set<String> deletedSmithyFiles = new HashSet<>(params.getChanges().size());
-        boolean changedBuildFiles = false;
-        for (FileEvent event : params.getChanges()) {
-            String changedUri = event.getUri();
-            if (changedUri.endsWith(".smithy")) {
-                if (event.getType().equals(FileChangeType.Created)) {
-                    createdSmithyFiles.add(changedUri);
-                } else if (event.getType().equals(FileChangeType.Deleted)) {
-                    deletedSmithyFiles.add(changedUri);
-                }
-            } else if (changedUri.endsWith(ProjectConfigLoader.SMITHY_BUILD)
-                    || changedUri.endsWith(ProjectConfigLoader.SMITHY_PROJECT)) {
-                changedBuildFiles = true;
-            } else {
-                for (String extFile : ProjectConfigLoader.SMITHY_BUILD_EXTS) {
-                    if (changedUri.endsWith(extFile)) {
-                        changedBuildFiles = true;
-                        break;
-                    }
-                }
+
+        Map<String, ProjectChanges> changesByProject = projects.computeProjectChanges(params.getChanges());
+
+        changesByProject.forEach((projectName, projectChanges) -> {
+            Project project = projects.getProjectByName(projectName);
+            if (projectChanges.hasChangedBuildFiles()) {
+                client.info("Build files changed, reloading project");
+                // TODO: Handle more granular updates to build files.
```

Comment:
Q: Could you provide me an example of what more granular updates you are thinking here?